### ISSUE TITLE
WISH: Built-in MD5 checksum calculator, e.g. `tools::md5(x)`

### ISSUE THREAD

**Issue by HenrikBengtsson** [HenrikBengtsson/Wishlist-for-R#21]
# Background
R has a built-in MD5 checksum calculator `tools::md5sum()`, but it only operates on files.  It takes a vector of pathnames (not connections) as input and returns a character string of the same length containing MD5 checksums, e.g.
```r
> files <- dir(R.home(), pattern = "^C", full.names = TRUE)
> files
[1] "C:/PROGRA~1/R/R-3.2.5/CHANGES" "C:/PROGRA~1/R/R-3.2.5/COPYING"
> tools::md5sum(files)
     C:/PROGRA~1/R/R-3.2.5/CHANGES      C:/PROGRA~1/R/R-3.2.5/COPYING
"d45eec95ce49830fdd3277950397dbde" "0cce1e42ef3fb133940946534fcf8896"
```

# Wish / Suggestion
Calculating MD5 checksums is such a common task that it would warrant a core R functions for calculating the checksum for an R object `x`, e.g. `tools::md5(x)`.

There is an internal [`src/library/tools/src/md5.c`](https://github.com/wch/r-source/blob/e5b21d0397c607883ff25cca379687b86933d730/src/library/tools/src/md5.c) file that implements the MD5 checksum.  It even has an internal [`md5_buffer()`](https://github.com/wch/r-source/blob/e5b21d0397c607883ff25cca379687b86933d730/src/library/tools/src/md5.c#L167-L184) function that seems to do exactly this.

# See also
* [digest](https://cran.r-project.org/package=digest) package:  Provides well-tested function `digest::digest(x, algo="md5")` for calculating the MD5 checksum for R object `x`.

**Comment by luciorq** [comment]
I would love to see an update on that front, is there any way to ping R core members responsible for `utils`?

**Comment by tdeenes** [comment]
yup, I tried it ~2.5 years ago on r-devel without success... See the convo [here](https://stat.ethz.ch/pipermail/r-devel/2020-May/079414.html)

**Comment by yihui** [comment]
FYI I asked an R core member in 2019 offline and he said it was not straightforward. Last week I had a precious chance to meet four R core members, and the good news is the larger sample size worked: https://github.com/wch/r-source/commit/c91b84567c8488b4bd3da6c184af05a9fc6e5e97 Hopefully we will be able to use `tools::md5sum(bytes = raw())` in the next version of R, which means we can `serialize()` an object to a `NULL` connection, get the raw bytes, pass them to `md5sum()`, and obtain the MD5 checksum.

**Comment by yihui** [comment]
Again FYI, here comes `sha256sum()`: https://github.com/wch/r-source/commit/a7c34a7ef40a53b2b0b7b9a47f322220de05aff7

**Comment by HenrikBengtsson** [comment]
These are excellent news. Now we just have to wait for R 4.5.0 to be released. I think we can close this wish.
